boat_simulator/nodes/data_collection/data_collection_node.py

```diff
@@ -0,0 +1,222 @@
+#!/usr/bin/env python3
+
+"""The ROS node for data collection."""
+import inspect
+import json
+import os
+import signal
+import sys
+from typing import Type
+
+import custom_interfaces.msg
+import rclpy
+import rclpy.utilities
+import rosbag2_py
+import rosidl_runtime_py
+from rclpy.node import Node
+from rclpy.serialization import serialize_message
+
+import boat_simulator.common.constants as Constants
+
+
+def shutdown_handler(signum, frame):
+    """Necessary for ros shutdown callback to be properly called."""
+    sys.exit(0)
+
+
+def main(args=None):
+    rclpy.init(args=args)
+    node = DataCollectionNode()
+    if is_collection_enabled():
+        try:
+            signal.signal(signal.SIGINT, shutdown_handler)
+            rclpy.spin(node)
+        finally:
+            rclpy.shutdown()
+
+
+def is_collection_enabled():
+    try:
+        is_data_collection_enabled_index = (
+            sys.argv.index(Constants.DATA_COLLECTION_CLI_ARG_NAME) + 1
+        )
+        is_data_collection_enabled = sys.argv[is_data_collection_enabled_index] == "true"
+    except ValueError:
+        is_data_collection_enabled = False
+    return is_data_collection_enabled
+
+
+class DataCollectionNode(Node):
+    def __init__(self):
+        super().__init__(node_name="data_collection_node")
+        self.get_logger().debug("Initializing node...")
+        self.__declare_ros_parameters()
+        self.__init_storage_files()
+        self.__init_msg_types_dict()
+        self.__init_subscriptions()
+        self.__init_timer_callbacks()
+        self.__init_shutdown_callbacks()
+        self.get_logger().debug("Node initialization complete. Starting execution...")
+
+    def __declare_ros_parameters(self):
+        """Declares ROS parameters from the global configuration file that will be used in this
+        node.
+        """
+        self.get_logger().debug("Declaring ROS parameters...")
+
+        self.declare_parameters(
+            namespace="",
+            parameters=[
+                ("file_name", rclpy.Parameter.Type.STRING),
+                ("qos_depth", rclpy.Parameter.Type.INTEGER),
+                ("topics", rclpy.Parameter.Type.STRING_ARRAY),
+                ("bag", rclpy.Parameter.Type.BOOL),
+                ("json", rclpy.Parameter.Type.BOOL),
+                ("write_period_sec", rclpy.Parameter.Type.DOUBLE),
+            ],
+        ),
+        all_parameters = self._parameters
+        for name, parameter in all_parameters.items():
+            value_str = str(parameter.value)
+            self.get_logger().debug(f"Got parameter {name} with value {value_str}")
+
+    def __init_storage_files(self):
+        """Initializes storage files for data logging."""
+        self.get_logger().debug("Initializing storrage files...")
+
+        if self.use_json:
```

Comment:
Consider this if statement and the next as their own helper functions and them call them here